### ISSUE TITLE
How do I read my data sent from php curl

### ISSUE THREAD

**Issue by Asimov500** [oktal/pistache#563]
Our company uses various apis at the moment which we run from php. However we need some more speed for some of our apis and C++ is a lot faster of course. So I have worked out how to get the routes working with Pastiche, and here is my code.

```
#include <pistache/router.h>
#include "pistache/endpoint.h"
#include "pistache/http.h"
#include <iostream>

using namespace Pistache;
using namespace Rest;

Rest::Router router;

void sausage(const Rest::Request& request, Http::ResponseWriter response){
    std::cout << "We have contact" << std::endl;
    response.send(Pistache::Http::Code::Ok, "Bottoms Up\n");
}

void about(const Rest::Request& request, Http::ResponseWriter response){
    std::cout << "Kettle" << std::endl;
    response.send(Pistache::Http::Code::Ok, "Can't hear you when I have the <b>kettle</b> on\n");
}

int main(){
    Routes::Get(router,"/ready",Routes::bind(&sausage));
    Routes::Post(router,"/about",Routes::bind(&about));
    Pistache::Address addr(Pistache::Ipv4::any(), Pistache::Port(9080));
    auto opts = Pistache::Http::Endpoint::options()
        .threads(10).flags(
        Pistache::Tcp::Options::ReuseAddr);
    Http::Endpoint server(addr);
    server.init(opts);
    server.setHandler(router.handler());
    server.serve();
    return 0;
}
```

Now this is working when I send using CURL in php, but the problem is that I do not know how to get the data I am sending. I will give an example of the format of the data I am sending.
```
$postData = http_build_query(
			array(
				'dstdata' => 'hello',
				'more' => 'bomb',
			)
		);
$ch = curl_init();
curl_setopt($ch, CURLOPT_URL, 'localhost:9999/about');
		curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
		curl_setopt($ch, CURLOPT_POSTFIELDS, $postData);
		curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
		$output = curl_exec($ch);
		print_r($output);		
		curl_close($ch);
		$response = (array)json_decode($output);		
		return $response;
```
My question is how do I grab the data that is being sent from curl. Bear in mind this could be just ordinary data like you see, but sometimes we send binary files too.

**Comment by Asimov500** [comment]
I think I figured it out. Instead of
`void about(const Rest::Request& request, Http::ResponseWriter response){`
I have to put
```
void about(const Http::Request& request, Http::ResponseWriter response){`
std::cout request.body() << std::endl;
```
Now all I have to do is to work how to get a binary file also, in the same array

**Comment by hydratim** [comment]
Take a look at https://github.com/oktal/pistache/blob/master/examples/rest_server.cc

**Comment by Asimov500** [comment]
Thanks for replying but request.param didn't work with the data I am sending, which is really a json string. However request.body() got my json string to the server. Now all I have to do is to convert the json string into either an array or something else so I can access it. The json string I send from curl could have a number of paremeters, and one could be a binary file also. So I am still trying to work this out.

**Comment by hydratim** [comment]
for json parsing I personally use https://github.com/nlohmann/json which I find to be quite flexible and easy to use with pistache.

**Comment by Asimov500** [comment]
Hi hydratim,
Strangely enough I just tried nlohmann and installed it fine, however it doesn't seem to work on my string which  comes through like this `dstdata=hello&more=bomb`.

I tried
```
auto name = request.body();
auto j3 = json::parse(name);
```
However it gives an error` parse error - unexpected 'd'` so I am suspecting that my string that is being received isn't proper json. The way I send to the api is the same way we send to our php apis.
The reason I need to use c++ for this project is that php is just not fast enough, so I have re-written the code to convert a DST to an svg in C++. That part works. Just need to get the server part working now.

**Comment by Asimov500** [comment]
I have solved this particular problem by posting in a different way
$data = [
			"name" => "Hagrid",
			"age" => "36"
		];                                                                    
$data_string = json_encode($data);                                                                                   

$ch = curl_init('localhost:9999/testbinary');                                                                      
curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");                                                                     
curl_setopt($ch, CURLOPT_POSTFIELDS, $data_string);                                                                  
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);                                                                      
curl_setopt($ch, CURLOPT_HTTPHEADER, array(                                                                          
    'Content-Type: application/json',                                                                                
    'Content-Length: ' . strlen($data_string))                                                                       
);                                                                                                                    
$result = curl_exec($ch);
print_r($result);

And I have used the nlohmann json library to read it and here is the function I use to return the data. This is only test data at the moment
```
void newtest(const Rest::Request& request, Http::ResponseWriter response){
    std::cout << "NewTest" << std::endl;
    auto jj = json::parse(request.body());
    std::string name = jj["name"];
    std::string age =jj["age"];
    std::string test = name+" is my name, and my age is "+age;
    response.send(Pistache::Http::Code::Ok, test);
}
```
I have got another problem but I will make a separate post for that. I wish there was some place to post where I am not seeming to have an issue, but to just ask questions

Anyway I hope this helps someone else.
